File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/taskqueue/TaskConsumer.java
package com.example.accountservice.taskqueue;

import static net.logstash.logback.argument.StructuredArguments.kv;

import com.example.accountservice.config.SqsProperties;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.SmartLifecycle;
import org.springframework.stereotype.Component;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.DeleteMessageRequest;
import software.amazon.awssdk.services.sqs.model.Message;
import software.amazon.awssdk.services.sqs.model.MessageAttributeValue;
import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;

/**
 * Waits on the Task Queue (SQS FIFO) via long polling, and when a message arrives, looks up the
 * handler by {@code taskType} (MessageAttributes) among the injected {@link TaskHandler} list (each
 * domain's Task Controller) and invokes it — this repeats the same structure as {@code
 * outbox/OutboxConsumer}, but for the Task Queue. A Domain/Integration Event (a fact — "X
 * happened") and a Task (a command — "do X") are conceptually different, so they are kept on
 * separate Consumers/queues (see scheduling.md, domain-events.md).
 *
 * <p>Handler succeeds → delete the message (ack). Handler fails (or no handler is registered) → do
 * not delete — once SQS's visibility timeout elapses, it's automatically redelivered
 * (at-least-once). Each Task Controller/Command Service's idempotency assumes this redelivery (see
 * the 3 levels of idempotency in scheduling.md).
 *
 * <p>The background loop design has the same rationale as {@code outbox/OutboxConsumer} — hanging a
 * blocking loop that repeats long polling (up to waitTimeSeconds) on the {@code @Scheduled} thread
 * pool risks delaying other scheduled tasks, so start/stop is managed via a dedicated single-thread
 * {@link ExecutorService} + {@link SmartLifecycle}.
 */
@Component
public class TaskConsumer implements SmartLifecycle {

    private static final Logger log = LoggerFactory.getLogger(TaskConsumer.class);
    private static final String TASK_TYPE_ATTRIBUTE = "taskType";

    private final SqsClient sqsClient;
    private final SqsProperties sqsProperties;
    private final Map<String, TaskHandler> handlers;
    private final ExecutorService executor =
            Executors.newSingleThreadExecutor(r -> new Thread(r, "task-consumer"));

    private volatile boolean running = false;

    public TaskConsumer(
            SqsClient sqsClient, SqsProperties sqsProperties, List<TaskHandler> taskHandlers) {
        this.sqsClient = sqsClient;
        this.sqsProperties = sqsProperties;
        this.handlers =
                taskHandlers.stream()
                        .collect(Collectors.toMap(TaskHandler::taskType, Function.identity()));
    }

    @Override
    public void start() {
        running = true;
        executor.submit(this::pollLoop);
    }

    @Override
    public void stop() {
        running = false;
        executor.shutdown();
        try {
            if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
                executor.shutdownNow();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            executor.shutdownNow();
        }
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    private void pollLoop() {
        String queueUrl = sqsProperties.taskQueueUrl();
        while (running) {
            try {
                ReceiveMessageResponse response =
                        sqsClient.receiveMessage(
                                ReceiveMessageRequest.builder()
                                        .queueUrl(queueUrl)
                                        .maxNumberOfMessages(10)
                                        .messageAttributeNames(TASK_TYPE_ATTRIBUTE)
                                        .waitTimeSeconds(5)
                                        .build());
                for (Message message : response.messages()) {
                    handleMessage(queueUrl, message);
                }
            } catch (Exception e) {
                if (running) {
                    log.error("Failed to receive from Task Queue", e);
                    sleepBeforeRetry();
                }
            }
        }
    }

    private void handleMessage(String queueUrl, Message message) {
        MessageAttributeValue attribute = message.messageAttributes().get(TASK_TYPE_ATTRIBUTE);
        String taskType = attribute != null ? attribute.stringValue() : null;
        try {
            if (taskType == null) {
                throw new IllegalStateException("Missing taskType message attribute.");
            }
            TaskHandler handler = handlers.get(taskType);
            if (handler == null) {
                throw new IllegalStateException("No Task Handler registered for: " + taskType);
            }
            handler.handle(message.body());
            sqsClient.deleteMessage(
                    DeleteMessageRequest.builder()
                            .queueUrl(queueUrl)
                            .receiptHandle(message.receiptHandle())
                            .build());
        } catch (Exception e) {
            log.error("Failed to process task", kv("task_type", taskType), e);
            // Do not delete — it will be received again and retried after the visibility timeout.
        }
    }

    private void sleepBeforeRetry() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: implementations/fastapi/examples/src/task_queue/task_consumer.py
from __future__ import annotations

import asyncio
import json
import logging

import aioboto3
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from ..config.aws_config import AwsConfig
from ..config.sqs_config import SqsConfig
from .task_handlers import build_task_handlers

logger = logging.getLogger(__name__)


class TaskConsumer:
    """Waits to receive from the Task queue via long polling, and once a message is
    received, looks up the TaskController method in the dict `build_task_handlers()`
    assembled by `taskType` (MessageAttributes) and calls it — exactly the same structure
    as the Domain Event's `OutboxConsumer` (src/outbox/outbox_consumer.py).

    Since a TaskController lets exceptions propagate as-is (scheduling.md), what needs to
    happen here is the same as OutboxConsumer: success → `delete_message` (ack), failure
    (or no registered handler) → not deleted. Once the FIFO queue's visibility timeout
    passes, it's automatically redelivered, and once `maxReceiveCount` is exceeded, it's
    isolated into the DLQ (see "DLQ monitoring" in scheduling.md).
    """

    def __init__(self, session_factory: async_sessionmaker[AsyncSession]) -> None:
        self._session_factory = session_factory
        self._boto_session = aioboto3.Session()

    async def run_forever(self) -> None:
        queue_url = SqsConfig().task_queue_url  # type: ignore[call-arg]
        async with self._boto_session.client("sqs", **AwsConfig().client_kwargs()) as sqs_client:  # type: ignore[call-arg]
            while True:
                try:
                    result = await sqs_client.receive_message(
                        QueueUrl=queue_url,
                        MaxNumberOfMessages=10,
                        MessageAttributeNames=["taskType"],
                        WaitTimeSeconds=5,
                    )
                except asyncio.CancelledError:
                    raise
                except Exception:  # noqa: BLE001 - the receive loop must never die from an exception
                    logger.exception("Failed to receive Task from SQS")
                    await asyncio.sleep(1)
                    continue

                for message in result.get("Messages", []):
                    await self._handle_message(sqs_client, queue_url, message)

    async def _handle_message(self, sqs_client, queue_url: str, message: dict) -> None:
        task_type = message.get("MessageAttributes", {}).get("taskType", {}).get("StringValue")
        try:
            if not task_type:
                raise ValueError("The taskType message attribute is missing.")

            payload = json.loads(message.get("Body", "{}"))
            async with self._session_factory() as session:
                handler = build_task_handlers(session).get(task_type)
                if handler is None:
                    raise ValueError(f"No registered Task handler: {task_type}")
                await handler(payload)
                await session.commit()

            await sqs_client.delete_message(QueueUrl=queue_url, ReceiptHandle=message["ReceiptHandle"])
        except asyncio.CancelledError:
            raise
        except Exception:  # noqa: BLE001 - it must not be deleted, so it's retried after the visibility timeout
            logger.exception("Failed to process Task: task_type=%s", task_type)
